keps/sig-node/4817-resource-claim-device-status/README.md

```diff
@@ -0,0 +1,614 @@
+* Establish a standardization for device status (e.g. Network Devices status).
+* Enable 3rd party implementation of new functionalities based on the Device
+  Status (e.g. Secondary Network Services if the IPs of a network device are
+reported).
+
+### Non-Goals
+
+* Implement new functionalities based on the Device Status (e.g. Kubernetes
+  EndpointSlice Controller supporting IPs reported by networking DRA drivers).
+* Modifying the Resource Claim workflow to include the device status report.
+
+## Proposal
+
+### API - ResourceClaim.Status
+
+The API changes define a new `DeviceStatuses` field in the existing
+`ResourceClaimStatus` struct. `DeviceStatuses` is a slice of a new struct
+`AllocatedDeviceStatus` which holds device specific information. 
+
+A device, identified by `<driver name>/<pool name>/<device name>` can be
+represented only once in the `DeviceStatuses` slice and will also mention which
+request caused the device to be allocated. The state and characteristics of the
+device are reported in the `Conditions`, representing the operational state of
+the device and in the `DeviceInfo`, an arbitrary data slice representing device
+specific characteristics. Additionally, for networking devices, a field
+`NetworkDeviceInfo` can be used to report the IPs, the MAC address and the
+interface name.
+
+`DeviceInfo` being a slice of arbitrary data allows the DRA Driver to store
+device specific data in different formats. For example, a Network Device being
+configured by via a CNI plugin could get its `DeviceInfo` filled with the CNI
+result for troubleshooting purpose and with a Network result in a modern
+standard format (closer to Pod.Status.PodIPs for example) used by 3rd party
+controllers.
+
+For each device, if required, the DRA Driver processing the device allocation
+can report the status of it in the `Status.DeviceStatuses` of the ResourceClaim
+by using the Kubernetes API.
+
+```golang
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the result of that was.
+type ResourceClaimStatus struct {
+	...
+
+	// DeviceStatuses contains the status of each device allocated for this
+	// claim, as reported by the driver. This can include driver-specific
+	// information. Entries are owned by their respective drivers.
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=devicePoolName
+	// +listMapKey=deviceName
+	DeviceStatuses []AllocatedDeviceStatus `json:"deviceStatuses,omitempty" protobuf:"bytes,4,opt,name=deviceStatuses"`
+}
+
+
+// AllocatedDeviceStatus contains the status of an allocated device, if the
+// driver chooses to report it. This may include driver-specific information.
+type AllocatedDeviceStatus struct {
+	// Request is the name of the request in the claim which caused this
+	// device to be allocated. Multiple devices may have been allocated
+	// per request.
+	//
+	// +required
+	Request string `json:"request" protobuf:"bytes,1,rep,name=request"`
```

Comment:
Is this needed? More generally, can we drop the "Allocated" part from the API?

It is not useful to post status of arbitrary, unrelated devices, but it can happen at least temporarily, for example directly after a claim got deallocated.

Speaking of which, who removes such stale entries?

A claim can go from "new" to "allocated" back to "deallocated" and then to "allocated" again.

Reply:
Right, I removed it, I guess this is repetitive as this information can be retrieved from `.Status.Allocation.Devices.Results`.

In the `Notes/Constraints/Caveats` section, I stated `The accuracy of the information depends on the implementation of the DRA Drivers, the reported status of the device may not always reflect the real time changes of the state of the device.` so the DRA-Driver should update these information based on when it configures/check/unconfigured devices.

Probably if the `ResourceClaim` gets deallocated then the `DeviceStatuses` can be cleared as it might be deleted or reallocated in which case the driver will have to reconfigure the devices in the pods/containers?

Reply:
I suppose that could be added. In that case, calling this `AllocatedDeviceStatus` would make sense because it is meant and documented as "only for allocated devices".